### PR TITLE
Include/Exclude support

### PR DESCRIPTION
**Pull request by emgarten**
Include/exclude support for projet.json
https://github.com/NuGet/Home/wiki/%5BSpec%5D-Managing-dependency-package-assets

//cc @davidfowl @yishaigalatzer @deepakaravindr @MeniZalzman @feiling

### CODE REVIEW

**Comment by emgarten** [comment]
//cc @anurse

File: src/NuGet.Core/NuGet.Commands/IncludeFlagUtils.cs [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+            var directDependencies = new SortedSet<string>(
+                spec.Dependencies.Select(dependency => dependency.Name),
+                StringComparer.OrdinalIgnoreCase);
+
+            var unifiedNodes = new Dictionary<string, GraphItem<RemoteResolveResult>>(StringComparer.OrdinalIgnoreCase);
+
+            // Create a look up table of id -> library
+            // This should contain only packages and projects. If there is a project with the 
+            // same name as a package, use the project.
```
Comment (MeniZalzman) [inline src/NuGet.Core/NuGet.Commands/IncludeFlagUtils.cs:31]: Please comment why we are doing the orderby

**Comment by analogrelay** [comment]
I don't know that I can review every line because it's quite a bit of code, but I gave it a look over and nothing looks horribly scary. It looks pretty reasonable though. As long as it's got some tests, I'd be OK with signing off.

File: src/NuGet.Core/NuGet.LibraryModel/LibraryIncludeType.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+using System.Linq;
+using NuGet.Common;
+
+namespace NuGet.LibraryModel
+{
+    public class LibraryIncludeType : IEquatable<LibraryIncludeType>
+    {
+        private readonly HashSet<LibraryIncludeTypeFlag> _keywords;
+
+        public static LibraryIncludeType Default = new LibraryIncludeType(LibraryIncludeTypeKeyword.Default.FlagsToAdd);
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.LibraryModel/LibraryIncludeType.cs:15]: why these are not readonly?

Can we put all the publics together

File: src/NuGet.Core/NuGet.Commands/RestoreCommand.cs [hunk 3]
```diff
@@ -434,56 +434,69 @@ private bool ResolutionSucceeded(IEnumerable<RestoreTargetGraph> graphs)
...
+                            targets.AddRange(items
+                                .Select(c => c.Path)
+                                .Where(path => Path.GetExtension(path).Equals(".targets", StringComparison.OrdinalIgnoreCase))
+                                .Select(path =>
+                                    Path.Combine(pathResolver.GetPackageDirectory(library.Key.Name, library.Key.Version),
+                                    path.Replace('/', Path.DirectorySeparatorChar))));
+
+                            props.AddRange(items
+                                .Select(c => c.Path)
+                                .Where(path => Path.GetExtension(path).Equals(".props", StringComparison.OrdinalIgnoreCase))
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand.cs:495]: you should have a single where, and a single select here.

**Comment by MeniZalzman** [comment]
:shipit:

File: src/NuGet.Core/NuGet.Commands/IncludeFlagUtils.cs [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+                else
+                {
+                    result.Add(dependency.Name, dependency.IncludeType);
+                }
+            }
+
+            return result;
+        }
+
+        private static void FlattenDependencyTypesUnified(
```
Comment (davidfowl) [inline src/NuGet.Core/NuGet.Commands/IncludeFlagUtils.cs:78]: Wow...

**Pull request closed without merging**